File: hafman.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

// HNode - узел дерева Хаффмана
type HNode struct {
	symbol     byte
	probability float64
	code       string
	parent     *HNode
	left       *HNode
	right      *HNode
}

// NewHNodeFromPair - конструктор из пары (вероятность, символ)
func NewHNodeFromPair(sym struct {
	first  float64
	second byte
}) *HNode {
	return &HNode{
		symbol:      sym.second,
		probability: sym.first,
		code:        "",
		parent:      nil,
		left:        nil,
		right:       nil,
	}
}

// NewHNodeFromChildren - конструктор из двух дочерних узлов
func NewHNodeFromChildren(l, r *HNode) *HNode {
	return &HNode{
		symbol:      '\x00',
		probability: l.probability + r.probability,
		code:        "",
		parent:      nil,
		left:        l,
		right:       r,
	}
}

// CharStringPair - аналог pair<char, string>
type CharStringPair struct {
	first  byte
	second string
}

// DoubleCharPair - аналог pair<double, char>
type DoubleCharPair struct {
	first  float64
	second byte
}

// String - реализация вывода для CharStringPair
func (p CharStringPair) String() string {
	return fmt.Sprintf("%c %s", p.first, p.second)
}

// String - реализация вывода для DoubleCharPair
func (p DoubleCharPair) String() string {
	return fmt.Sprintf("%f %c", p.first, p.second)
}

// createCode - функция для создания кодов
func createCode(node *HNode, currentCode string) {
	if node == nil {
		return
	}

	node.code = currentCode

	if node.left != nil {
		createCode(node.left, currentCode+"0")
	}
	if node.right != nil {
		createCode(node.right, currentCode+"1")
	}
}

// buildCodeTable - построение таблицы кодов
func buildCodeTable(codeTable []CharStringPair, currentNode *HNode, index *int) {
	if currentNode == nil {
		return
	}
	if currentNode.symbol == '\x00' {
		buildCodeTable(codeTable, currentNode.left, index)
		buildCodeTable(codeTable, currentNode.right, index)
		return
	}
	codeTable[*index] = CharStringPair{currentNode.symbol, currentNode.code}
	*index++
}

// HafmanCode - структура кода Хаффмана
type HafmanCode struct {
	codeTable []CharStringPair
	size      int
}

// NewHafmanCode - конструктор кода Хаффмана
func NewHafmanCode(str string) *HafmanCode {
	// Получаем уникальные символы
	unique := make([]byte, 0)
	for _, s := range str {
		found := false
		for _, u := range unique {
			if u == byte(s) {
				found = true
				break
			}
		}
		if !found {
			unique = append(unique, byte(s))
		}
	}

	// Создаем таблицу вероятностей
	table := make([]DoubleCharPair, len(unique))
	for i, char := range unique {
		count := 0
		for _, s := range str {
			if byte(s) == char {
				count++
			}
		}
		table[i] = DoubleCharPair{
			first:  float64(count) / float64(len(str)),
			second: char,
		}
	}

	// Сортируем по вероятности
	sort.Slice(table, func(i, j int) bool {
		return table[i].first < table[j].first
	})

	// Создаем рабочую таблицу узлов
	workTable := make([]*HNode, len(unique))
	for i := 0; i < len(unique); i++ {
		workTable[i] = NewHNodeFromPair(table[i])
	}

	sizeWorkTable := len(unique)
	for sizeWorkTable != 1 {
		// Создаем новый узел из двух наименьших
		node := NewHNodeFromChildren(workTable[0], workTable[1])
		workTable[0].parent = node
		workTable[1].parent = node
		sizeWorkTable--

		// Сдвигаем элементы
		for i := 0; i < sizeWorkTable; i++ {
			workTable[i] = workTable[i+1]
		}
		workTable[0] = node

		// Сортируем
		sort.Slice(workTable[:sizeWorkTable], func(i, j int) bool {
			return workTable[i].probability < workTable[j].probability
		})
	}

	head := workTable[0]
	createCode(head, "")

	// Создаем таблицу кодов
	codeTable := make([]CharStringPair, len(unique))
	index := 0
	buildCodeTable(codeTable, head, &index)

	result := ""
	for _, char := range str {
		for j := 0; j < len(unique); j++ {
			if codeTable[j].first == byte(char) {
				result += codeTable[j].second
			}
		}
	}
	fmt.Printf("Код: %s\n", result)
	fmt.Println("Таблица кодирования:")
	for i := 0; i < len(unique); i++ {
		fmt.Println(codeTable[i])
	}

	return &HafmanCode{
		codeTable: codeTable,
		size:      len(unique),
	}
}

// decode - декодирование строки
func (h *HafmanCode) decode(code string) {
	result := ""
	decodeSymbol := ""
	for i := 0; i < len(code); i++ {
		decodeSymbol += string(code[i])
		for j := 0; j < h.size; j++ {
			if h.codeTable[j].second == decodeSymbol {
				result += string(h.codeTable[j].first)
				decodeSymbol = ""
			}
		}
	}
	fmt.Printf("Декодированная строка: %s\n", result)
}

func hafman() {
	fmt.Println("Введите строку для кодирования:")

	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	input := scanner.Text()

	huffman := NewHafmanCode(input)

	fmt.Println("\nВведите код для декодирования (или 'stop' для выхода):")
	scanner.Scan()
	codeToDecode := scanner.Text()

	for codeToDecode != "stop" {
		huffman.decode(codeToDecode)

		fmt.Println("\nВведите следующий код для декодирования (или 'stop' для выхода):")
		scanner.Scan()
		codeToDecode = scanner.Text()
	}
}

File: hafman.cpp
#include <iostream>
#include <algorithm>
#include <sstream>

using namespace std;

struct HNode{
    char symbol;
    double probability;
    string code;
    HNode* parent;
    HNode* left;
    HNode* right;
    HNode(pair<double, char> sym) : symbol(sym.second)
                                , probability(sym.first)
                                , code("")
                                , parent(nullptr)
                                , left(nullptr)
                                , right(nullptr){}
    HNode(HNode* l, HNode* r) : symbol('\0')
                                , probability(l -> probability + r->probability)
                                , code("")
                                , parent(nullptr)
                                , left(l)
                                , right(r){}
};

ostream& operator<<(ostream& ss, pair<char, string> dc){
    ss << dc.first << " " << dc.second;
    return ss;
}

ostream& operator<<(ostream& ss, pair<double, char> dc){
    ss << dc.first << " " << dc.second;
    return ss;
}

// Функция для создания кодов
void createCode(HNode* node, string currentCode = "" ){
    if (!node) return;
    
    node->code = currentCode;
    
    if (node->left != nullptr) createCode(node->left, currentCode + "0");
    if (node->right != nullptr) createCode(node->right, currentCode + "1");
}

void buildCodeTable(pair<char, string>* codeTable, HNode* curentNode, int& index){
    if (curentNode == nullptr) return;
    if (curentNode -> symbol == '\0'){
        buildCodeTable(codeTable, curentNode -> left, index);
        buildCodeTable(codeTable, curentNode -> right, index);
        return;
    }
    codeTable[index] = {curentNode -> symbol, curentNode -> code};
    index++;
}

struct HafmanCode{
    pair<char, string>* codeTable;
    size_t size;
    HafmanCode(string str){
        string unique;
        for (char s : str) if (unique.find(s) == string::npos) unique += s;

        pair<double, char> Table[unique.size()];
        int* j = new int(0);
        for (char i : unique){
            int cnt = count(str.begin(), str.end(), i);
            Table[(*j)++] = {static_cast<double>(cnt)/static_cast<double>(str.size()), i};
        }
        delete j;

        sort(Table, Table + unique.size(), [](pair<double, char> a, pair<double, char> b){
            return  a.first < b.first;
        });

        HNode* workTable[unique.size()];
        for(int i = 0; i < unique.size(); i++) {
            HNode* node = new HNode(Table[i]);
            workTable[i] = node;
        }

        size_t sizeWorkTable = unique.size();
        while(sizeWorkTable != 1){
            HNode* node = new HNode(workTable[0], workTable[1]);
            workTable[0] -> parent = node;
            workTable[1] -> parent = node;
            sizeWorkTable--;
            for (int i = 0; i < sizeWorkTable; i++) workTable[i] = workTable[i+1];
            workTable[0] = node;
            sort(workTable, workTable + sizeWorkTable, [](HNode* a, HNode* b){
                return  a -> probability < b -> probability;
            });
        }

        HNode* head;
        head = workTable[0];
        createCode(head);

        codeTable = new pair<char, string>[unique.size()];
        size = unique.size();
        int index = 0;
        buildCodeTable(codeTable, head, index);
        
        string result = "";
        for(char i : str){
            for(int j = 0; j < unique.size(); j++) if(codeTable[j].first == i) result += codeTable[j].second;
        }
        cout << "Код: " << result << endl;
        cout << "Таблица кодирования:" << endl;
        for(int i = 0; i < unique.size(); i++) cout << codeTable[i] << endl;
    }
    
    // Добавляем деструктор для очистки памяти
    ~HafmanCode() {
        delete[] codeTable;
    }
    
    void decode(string code){
        string result = "";
        string decodeSymbol = "";
        for (int i = 0; i < code.size(); i++){
            decodeSymbol += code[i];
            for(int j = 0; j < size; j++){
                if(codeTable[j].second == decodeSymbol) {
                    result += codeTable[j].first;
                    decodeSymbol = "";
                }
            } 
        }
        cout << "Декодированная строка: " << result << endl;
    }
};

void hafman(){
    cout << "Введите строку для кодирования:" << endl;
    
    string input;
    cin.ignore();
    getline(cin, input);
    HafmanCode huffman(input);

    cout << endl << "Введите код для декодирования (или 'stop' для выхода):" << endl;
    string code_to_decode;
    getline(cin, code_to_decode);
    
    while (code_to_decode != "stop") {
        huffman.decode(code_to_decode);
        
        cout << endl << "Введите следующий код для декодирования (или 'stop' для выхода):" << endl;
        getline(cin, code_to_decode);
    }
}